core/classes/IssueStatusChangeTimelineEvent.class.php

```diff
@@ -45,24 +55,67 @@ public function __construct( $p_timestamp, $p_user_id, $p_issue_id, $p_old_statu
 		$this->issue_id = $p_issue_id;
 		$this->old_status = $p_old_status;
 		$this->new_status = $p_new_status;
+		$this->type = $this->change_type();
 	}
 
 	/**
-	 * Returns html string to display
-	 * @return string
+	 * Return the type of status change
+	 * @return int One of the status change constants defined above
 	 */
-	public function html() {
+	private function change_type() {
 		$t_resolved = config_get( 'bug_resolved_status_threshold' );
 		$t_closed = config_get( 'bug_closed_status_threshold' );
 
-		if( $this->old_status < $t_closed && $this->new_status >= $t_closed ) {
-			$t_string = sprintf( lang_get( 'timeline_issue_closed' ), user_get_name( $this->user_id ), string_get_bug_view_link( $this->issue_id ) );
-		} else if( $this->old_status < $t_resolved && $this->new_status >= $t_resolved ) {
-			$t_string = sprintf( lang_get( 'timeline_issue_resolved' ), user_get_name( $this->user_id ), string_get_bug_view_link( $this->issue_id ) );
+		if( $this->old_status < $t_resolved && $this->new_status >= $t_resolved ) {
+			return IssueStatusChangeTimelineEvent::RESOLVED;
+		} else if( $this->old_status < $t_closed && $this->new_status >= $t_closed ) {
+			return IssueStatusChangeTimelineEvent::CLOSED;
 		} else if( $this->old_status >= $t_resolved && $this->new_status < $t_resolved ) {
-			$t_string = sprintf( lang_get( 'timeline_issue_reopened' ), user_get_name( $this->user_id ), string_get_bug_view_link( $this->issue_id ) );
+			return IssueStatusChangeTimelineEvent::REOPENED;
 		} else {
-			return '';
+			return IssueStatusChangeTimelineEvent::IGNORED;
+		}
+	}
+
+	/**
+	 * Whether to skip this timeline event.
+	 * This normally implements access checks for the event.
+	 * @return boolean
+	 */
+	public function skip() {
```

Comment:
My initial approach followed a similar but less elaborated solution; the filtering logic was in timeline_print_events() function, testing an empty string returned by html() within the foreach loop.

Then I thought it was simpler and more elegant to have a "clean" array (i.e. only containing useful data) to begin with.

> option of having an event that generates more than one row

I'm not sure what is the use for this, or if it's even possible. Did you have anything specific in mind ? 
I think that having possible multiple entries would make counting events much more complex, and I'm not sure it's worth it.

> we don't need to synchronize the logic in skip() and html()

They are used separately and don't need to be in sync: 
- skip() is used to exclude events from the array built by timeline_events() and 
- timeline_print_events() / html() just print whatever is in the array.

A possibly simpler alternative to reintroducing "skip" could be to have the constructor throw an exception.